nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/AppLibClassLoaderServiceImpl.java

```diff
@@ -17,165 +17,547 @@
 
-        AppLibClassFinder(ClassLoader parent, List<DelegatingClassLoader.ClassFinder> delegates)
-                throws IllegalArgumentException {
-            super(parent, delegates);
+        /**
+         * Returns an enumeration of URL object representing the resource with the given name.
+         *
+         * @param name the resource name
+         * @return a singleton enumeration of URL object for the resource, or empty enumeration
+         * if the resources not found
+         */
+        @Override
+        public Enumeration<URL> findResources(String name) {
+            URL resourceURL = findResource(name);
+            return resourceURL != null ? enumeration(List.of(resourceURL)) : emptyEnumeration();
         }
+    }
+
+    /**
+     * This class loader has a list of class loaders called as delegates
+     * that it uses to find resources and classes.
+     */
+    private static class DelegatingClassFinder extends DelegatingClassLoader implements ClassFinder {
 
-        AppLibClassFinder(ClassLoader parent) {
+        DelegatingClassFinder(ClassLoader parent) {
             super(parent);
         }
 
+        /**
+         * Always returns {@code null} because delegating class loader will
+         * never be a defining class loader.
+         */
         @Override
         public Class<?> findExistingClass(String name) {
-            // no action needed as parent is delegating classloader which will never be a defining classloader
             return null;
         }
 
+        /**
+         * Finds the resource with the given name.
+         *
+         * @param name the resource name
+         * @return a URL object for reading the resource, or {@code null} if the resource
+         * could not be found
+         */
         @Override
         public URL findResource(String name) {
             return super.findResource(name);
         }
 
+        /**
+         * Returns an enumeration of URL objects representing all resources with the given name.
+         *
+         * @param name the resource name
+         * @return an enumeration of URL objects for the resources, or empty enumeration
+         * if the resources not found
+         * @throws IOException if an I/O error occurs
+         */
         @Override
         public Enumeration<URL> findResources(String name) throws IOException {
             return super.findResources(name);
         }
     }
+
+    /**
+     * Represents a deployment time library.
+     */
+    private static class Library {
+
+        /**
+         * Represents a method to read the attributes of an open file.
+         */
+        private static final Method readAttributesMethod;
+
+        /**
+         * Represents a field to access to the native descriptor for an open file.
+         */
+        private static final Field nativeDescriptorField;
+
+        static {
+            Method method;
+            Field field;
+            try {
+                if (OS.isWindows()) {
```

Comment:
I am not sure now, but isn't there already a way how to get these with NIO? Maybe just with newer JDK versions, I am not sure now. I have to check it later ...

Reply:
> Isn't there already a way how to get these with NIO? Maybe just with newer JDK versions, I am not sure now. I have to check it later ...

I just need to add logging and do some code cleanup... After that, I will add a detailed description of why the decision was made to use certain platform-dependent means.

Reply:
The problem is that it may change or even be abandoned in the future. How critical is the difference from this? https://www.baeldung.com/java-nio2-file-attribute

Is this the whole explanation?

> If we are unable to create a temporary snapshot of the library using native descriptor, we fall back to using standard NIO.2 functions.

Reply:
> The problem is that it may change or even be abandoned in the future. How critical is the difference from this? https://www.baeldung.com/java-nio2-file-attribute

This prevents inconsistency between file attributes and library content in case where library has been replaced with new one after the attributes were read, but before the snapshot was created.

First, I open `FileInputStream`. This block the file from being replaced or deleted on Windows and saves the contents of the file (inode?) on UNIX until the input stream closes. Then I get the open file descriptor (handle on Windows) and read the attributes from it. Then I copy the open input stream to a temporary file and close input stream. This ensures that the attributes match the file exactly.

If for some reason it is impossible to read attributes using os-dependent functions, we return to using standard NIO.2 functions/

Reply:
> it may change or even be abandoned in the future

In this case snandard NIO.2 will be used. Maybe until another solution is found.

Reply:
I did test on Temurin 11-22 and some latest Semeru.

I also tried using simple tests how to get the attributes of open files on Windows, Linux, macOS and Solaris(OpenIndiana). For OpenJDK from 11 to 22 all works now.

Reply:
> It still smells to me as some bad practice

Unfortunately, Java doesn't have methods `FileInputStream.getAttibutes()`, `FileChannel.getAttributes()`, `Files.readAttributes(FileDescriptor fd, ...)` or similar. Therefore we are forced to use the "reflection magic".


> lets try that, if somebody finds it problematic in practice, I hope he will report it

This should never fail. If "reflection magic" does not work, the standard path will:

```java
            // Fallback to the standard NIO.2 method
            if (attributes == null) {
                attributes = readAttributes(libURI);
            }
```

In `readAttributes() method`

```java
            return Files.readAttributes(Path.of(libURI), BasicFileAttributes.class);
```